Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx

```diff
@@ -1,26 +1,128 @@
+import { useCallback, useReducer, useState } from "react";
+
 import BatteryInput from "~/components/batteryInput";
 import Header from "~/components/header";
 import SpeedInput from "~/components/speedInput";
 import WeatherInput from "~/components/weatherInput";
 
+interface inputState {
+  speed: number;
+  battery: number;
+  weather: number;
+}
+
+interface inputAction {
+  type: "changed_speed" | "changed_battery" | "changed_weather";
+  payload: number;
+}
+
 const App = () => {
-  const calculateRange = () => {
-    return;
-  }
+  const [showResult, setShowResult] = useState(false);
+  const [errors, setErrors] = useState({ speedError: "", batteryError: "" });
+
+  //the reducer function, updates the state when input values (speed,battery,weather) are changed
+  const reducer = (state: inputState, action: inputAction) => {
+    setShowResult(false); // Reset showResult when inputs change
+    switch (action.type) {
+      case "changed_speed":
+        return { ...state, speed: action.payload };
+      case "changed_battery":
+        return { ...state, battery: action.payload };
+      case "changed_weather":
+        return { ...state, weather: action.payload };
+      default:
+        return state;
+    }
+  };
+
+  const [state, dispatch] = useReducer(reducer, {
+    //useReducer to manage more complex states
+    speed: 0, //the initial state of speed, battery, weather values
+    battery: 0,
+    weather: 50,
+  });
+
+  //validate speed and battery inputs
+  const validateInputs = useCallback(() => {
```

Comment:
not sure if this logic is needed to be within a function, you can calculate the errors and memoize it